pkg/plugins/runtime/k8s/controllers/pod_status_controller.go

```diff
@@ -0,0 +1,121 @@
+package controllers
+
+import (
+	"context"
+
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	"github.com/kumahq/kuma/pkg/envoy/admin"
+	k8s_common "github.com/kumahq/kuma/pkg/plugins/common/k8s"
+	util_k8s "github.com/kumahq/kuma/pkg/plugins/runtime/k8s/util"
+
+	"github.com/go-logr/logr"
+	kube_core "k8s.io/api/core/v1"
+	kube_apierrs "k8s.io/apimachinery/pkg/api/errors"
+	kube_runtime "k8s.io/apimachinery/pkg/runtime"
+	kube_record "k8s.io/client-go/tools/record"
+	kube_ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/kumahq/kuma/pkg/core/resources/manager"
+	mesh_k8s "github.com/kumahq/kuma/pkg/plugins/resources/k8s/native/api/v1alpha1"
+)
+
+// PodStatusReconciler creates default resources for created Mesh
+type PodStatusReconciler struct {
+	kube_client.Client
+	kube_record.EventRecorder
+	Scheme            *kube_runtime.Scheme
+	ResourceManager   manager.ResourceManager
+	Log               logr.Logger
+	ResourceConverter k8s_common.Converter
+	EnvoyAdminClient  admin.EnvoyAdminClient
+}
+
+func (r *PodStatusReconciler) Reconcile(req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("pod-status", req.NamespacedName)
+
+	// Fetch the Pod instance
+	pod := &kube_core.Pod{}
+	if err := r.Get(ctx, req.NamespacedName, pod); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			return kube_ctrl.Result{}, nil
+		}
+		log.Error(err, "unable to fetch Pod")
+		return kube_ctrl.Result{}, err
+	}
+
+	hasSidecar := false
+	for _, cs := range pod.Status.ContainerStatuses {
+		if cs.Name == util_k8s.KumaSidecarContainerName {
+			hasSidecar = true
+			if cs.State.Terminated != nil {
+				// the sidecar already terminated
+				return kube_ctrl.Result{}, nil
+			}
+		} else if cs.State.Terminated == nil {
+			// at least one non-sidecar container not terminated
+			// no need to tell envoy to quit yet
+			return kube_ctrl.Result{}, nil
+		}
+	}
+
+	if hasSidecar {
+		log.V(1).Info("unterminated pod with sidecar found, sending quit to envoy")
+
+		dataplane := &mesh_k8s.Dataplane{}
+		if err := r.Get(ctx, req.NamespacedName, dataplane); err != nil {
+			if kube_apierrs.IsNotFound(err) {
+				log.V(1).Info("dataplane is not found", "name", req.NamespacedName)
+				return kube_ctrl.Result{}, nil
+			}
+			log.Error(err, "unable to fetch Dataplane")
+			return kube_ctrl.Result{}, err
+		}
+
+		dp := core_mesh.NewDataplaneResource()
+		if err := r.ResourceConverter.ToCoreResource(dataplane, dp); err != nil {
+			converterLog.Error(err, "failed to parse Dataplane", "dataplane", dataplane.Spec)
+			return kube_ctrl.Result{}, err
+		}
+
+		err := r.EnvoyAdminClient.PostQuit(dp)
+		if err != nil {
+			// We do not want the reconciler to ever fail. Just log the error
```

Comment:
why we do not want it to ever fail?

Reply:
Because, the reality is that sometimes it will fail, and that is not fatal but will create a flood of messages in the logs